### ISSUE TITLE
MongoDB authentication error "AuthenticationFailed: SCRAM-SHA-1 authentication failed, storedKey mismatch"

### ISSUE THREAD

**Issue by donovanmuller** [keptn/keptn#3936]
**Environment**
<!-- Please tell us about the environment in which this bug can be reproduced in -->

* Client OS (e.g., Linux, macOS, Windows): MacOS 11.3
* Keptn Version (`keptn version`): `0.8.2`
* Kubernetes Cloud Provider (e.g., GKE, AKS): EKS
* Kubernetes version (`kubectl version`): 

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.4-dirty", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"dirty", BuildDate:"2021-03-15T09:58:13Z", GoVersion:"go1.16.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.8-eks-96780e", GitCommit:"96780e1b30acbf0a52c38b6030d7853e575bcdf3", GitTreeState:"clean", BuildDate:"2021-03-10T21:32:29Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
```

<!-- Please also let us know about other components and their version if you believe that this issue is related to them (e.g., Istio, Web Browser Version). -->

**Affected Component**
<!-- Please tell us which component of Keptn is affected (leave empty if you are unsure). -->

* [ ] Docs
* [ ] CLI
* [ ] Kubernetes Integration
* [ ] Openshift Integration
* [ ] Helm
* [ ] Istio
* [ ] Bridge
* [ ] Approval
* [x] Datastore
* [ ] REST API
* [ ] eventbroker / distributor
* [ ] jmeter

**Describe the bug**
<!-- Please provide a clear and concise description of what the bug is. Code samples should be put in the **To Reproduce** section. -->

After running a Keptn install for a few days, all API requests are failing. If you look into the logs of the `shipyard-controller`, `mongodb-datastore` and `mongodb` you will see the following errors:

`shipyard-controller`
```
shipyard-controller-6844695949-ht4fd:shipyard-controller MongoDB client lost connection. Attempt reconnect.                                                                                                                                                                                                                        
shipyard-controller-6844695949-ht4fd:shipyard-controller Error retrieving projects from mongoDB: connection() : auth error: sasl conversation error: unable to authenticate using mechanism "SCRAM-SHA-1": (AuthenticationFailed) Authentication failed.                                                                           
shipyard-controller-6844695949-ht4fd:shipyard-controller [GIN] 2021/05/03 - 16:20:36 | 500 |   497.06822ms |   10.247.146.19 | GET      "/v1/event/triggered/sh.keptn.event.service.delete.finished"
```

`mongodb-datastore`
```
mongodb-datastore-84c975559-jdzgl:mongodb-datastore {"timestamp":"2021-05-03T16:26:54.310005181Z","logLevel":"DEBUG","message":"MongoDB client lost connection. Attempt reconnect."}
mongodb-datastore-84c975559-jdzgl:mongodb-datastore {"timestamp":"2021-05-03T16:26:54.330087199Z","logLevel":"ERROR","message":"error counting elements in events collection: connection() : auth error: sasl conversation error: unable to authenticate using mechanism \"SCRAM-SHA-1\": (AuthenticationFailed) Authentication failed."}
```

`mongodb`
```
mongodb-6b455c56fb-4lj56:2021-05-03T16:21:47.486+0000 I NETWORK  [conn1184792] received client metadata from 10.247.145.228:40042 conn: { driver: { name: "mongo-go-driver", version: "v1.4.6" }, os: { type: "linux", architecture: "amd64" }, platform: "go1.13.7" }                                                             
mongodb-6b455c56fb-4lj56:2021-05-03T16:21:47.493+0000 I ACCESS   [conn1184792] SCRAM-SHA-1 authentication failed for user on keptn from client 10.247.145.228:40042 ; AuthenticationFailed: SCRAM-SHA-1 authentication failed, storedKey mismatch                                                                                  
mongodb-6b455c56fb-4lj56:2021-05-03T16:21:47.493+0000 I NETWORK  [conn1184792] end connection 10.247.145.228:40042 (2271 connections now open)
```

**To Reproduce**
<!-- Please provide detailed instructions of how to reproduce the behaviour, including code samples if applicable. -->

1. Leave Keptn to run for a good amount of time. I have not noticed any specific time
2. Try access Bridge or check any service and you'll notice the API requests fail
3. Check logs and notice the `SCRAM-SHA-1 authentication failed, storedKey mismatch` error

**Additional context**
<!-- Add any other context about the problem here. -->

To restore the API, `kubectl delete ...` the Pods from the Deployment's above, `shipyard-controller` `mongodb-datastore` and `mongodb` and then everything is restored.

**Comment by christian-kreuzberger-dtx** [comment]
> Leave Keptn to run for a good amount of time. I have not noticed any specific time

Our Keptn installations are running since we released 0.8.2 without the aforementioned problem.

What looks suspicious to me is the following line:
```
mongodb-6b455c56fb-4lj56:2021-05-03T16:21:47.493+0000 I NETWORK  [conn1184792] end connection 10.247.145.228:40042 (2271 connections now open)
`` 
There should not be 2271 connections open, I think mongodb is running in some sort of limit here.

Could you please provide the following details:

1. How was Keptn installed (using `keptn install`, `helm`, or any other method)? Was it ever upgraded?
2. Did you ever modify the mongodb credentials?
3. Are there any 3rd-party Keptn-services running, and if so, in which version? (`kubectl -n keptn get deployments -owide`)

**Comment by donovanmuller** [comment]
1. Helm chart install. Control plane in one cluster, execution plane in another cluster. It has not been upgraded. Brand new install via ArgoCD Helm Application
2. In trying to debug the issue initially I did delete the credentials Secret, which was promptly replaced by ArgoCD and after restarting the services mentioned above, all was well (for a few days...)
3. There are no additional services running in the control plane, but the following in the execution plane:

Control plane:
```console
$ kubectl -n keptn get deployments -owide
NAME                    READY   UP-TO-DATE   AVAILABLE   AGE     CONTAINERS                        IMAGES                                                                        SELECTOR
api-gateway-nginx       2/2     2            2           7d15h   api-gateway-nginx                 docker.io/nginxinc/nginx-unprivileged:1.19.4-alpine                           app.kubernetes.io/instance=keptn,app.kubernetes.io/name=api-gateway-nginx
api-service             1/1     1            1           7d15h   api-service,distributor           docker.io/keptn/api:0.8.2,docker.io/keptn/distributor:0.8.2                   app.kubernetes.io/instance=keptn,app.kubernetes.io/name=api-service
approval-service        1/1     1            1           7d15h   approval-service,distributor      docker.io/keptn/approval-service:0.8.2,docker.io/keptn/distributor:0.8.2      app.kubernetes.io/instance=keptn,app.kubernetes.io/name=approval-service
bridge                  1/1     1            1           7d15h   bridge                            docker.io/keptn/bridge2:0.8.2                                                 app.kubernetes.io/instance=keptn,app.kubernetes.io/name=bridge
configuration-service   1/1     1            1           7d15h   configuration-service             docker.io/keptn/configuration-service:0.8.2                                   app.kubernetes.io/instance=keptn,app.kubernetes.io/name=configuration-service
lighthouse-service      1/1     1            1           7d15h   lighthouse-service,distributor    docker.io/keptn/lighthouse-service:0.8.2,docker.io/keptn/distributor:0.8.2    app.kubernetes.io/instance=keptn,app.kubernetes.io/name=lighthouse-service
mongodb                 1/1     1            1           5d20h   mongodb                           docker.io/centos/mongodb-36-centos7:1                                         app.kubernetes.io/instance=keptn,app.kubernetes.io/name=mongodb
mongodb-datastore       1/1     1            1           7d15h   mongodb-datastore,distributor     docker.io/keptn/mongodb-datastore:0.8.2,docker.io/keptn/distributor:0.8.2     app.kubernetes.io/instance=keptn,app.kubernetes.io/name=mongodb-datastore
remediation-service     1/1     1            1           7d15h   remediation-service,distributor   docker.io/keptn/remediation-service:0.8.2,docker.io/keptn/distributor:0.8.2   app.kubernetes.io/instance=keptn,app.kubernetes.io/name=remediation-service
secret-service          1/1     1            1           7d15h   secret-service                    docker.io/keptn/secret-service:0.8.2                                          app.kubernetes.io/instance=keptn,app.kubernetes.io/name=secret-service
shipyard-controller     1/1     1            1           7d15h   shipyard-controller,distributor   docker.io/keptn/shipyard-controller:0.8.2,docker.io/keptn/distributor:0.8.2   app.kubernetes.io/instance=keptn,app.kubernetes.io/name=shipyard-controller
statistics-service      1/1     1            1           7d15h   statistics-service,distributor    docker.io/keptn/statistics-service:0.8.2,docker.io/keptn/distributor:0.8.2    app.kubernetes.io/instance=keptn,app.kubernetes.io/name=statistics-service
```

Execution plane:
```console
$ kubectl -n keptn get deployments -owide
NAME                       READY   UP-TO-DATE   AVAILABLE   AGE     CONTAINERS                             IMAGES                                                                                 SELECTOR
argo-service               1/1     1            1           7d15h   distributor,argo-service               keptn/distributor:0.8.0,grabnerandi/argo-service:0.8.0                                 run=argo-service
generic-executor-service   1/1     1            1           14h     distributor,generic-executor-service   keptn/distributor:0.8.2,keptnsandbox/generic-executor-service:0.8.0                    run=generic-executor-service
keptn-helm-service         1/1     1            1           7d15h   helm-service,distributor               docker.io/keptn/helm-service:0.8.2-hotfix1,docker.io/keptn/distributor:0.8.2-hotfix1   app.kubernetes.io/instance=keptn,app.kubernetes.io/name=helm-service
keptn-jmeter-service       1/1     1            1           7d15h   distributor,jmeter-service             docker.io/keptn/distributor:0.8.2,docker.io/keptn/jmeter-service:0.8.2                 app.kubernetes.io/instance=keptn,app.kubernetes.io/name=jmeter-service
```

**Comment by donovanmuller** [comment]
This happened again and I'm leaning towards it being related to Keptn Pods shuffling around after a cluster scaling event (EKS, Spot EC2 nodes and Cluster Autoscaler). Note the Pod ages:

```console
$ kubectl get pods -n keptn
NAME                                    READY   STATUS    RESTARTS   AGE
api-gateway-nginx-5b854464cf-24wq6      1/1     Running   0          8m35s
api-gateway-nginx-5b854464cf-6k8mb      1/1     Running   0          8m32s
api-service-7bc6ff97fc-284c2            2/2     Running   0          24m
approval-service-dcb7979c8-wcms7        2/2     Running   0          24m
bridge-788f564489-fvpwk                 1/1     Running   0          24m
configuration-service-5bc9548bd-pp5n4   1/1     Running   0          140m
keptn-nats-cluster-0                    3/3     Running   0          10m
lighthouse-service-857df5646b-hbjkk     2/2     Running   0          124m
mongodb-6b455c56fb-c65qs                1/1     Running   0          140m
mongodb-datastore-84c975559-bqg4t       2/2     Running   0          124m
remediation-service-5b9ffb54f9-mjl6s    2/2     Running   0          26m
secret-service-7ffdf597d7-n67q7         1/1     Running   0          24m
shipyard-controller-6844695949-6brkr    2/2     Running   0          24m
statistics-service-7bd9968dbd-p6mlz     2/2     Running   0          26m
```

I'll check next time this happens but could it be that the `mongodb` Pod get's killed and spun up on another node and the services, in an attempt to reconnect cause some sort of issue that can only be resolved by forcing those services to restart only once `mongodb` is actually running again? 🤔

**Comment by christian-kreuzberger-dtx** [comment]
If that's the case we should be able to reproduce it by killing mongodb on purpose.

**Comment by christian-kreuzberger-dtx** [comment]
I tried killing the mongodb pod and see what happens. So far nothing spectacular, everything is still working after the pod came back up.

What I am still puzzled about is the fact that your mongodb is telling that there are 2271 open connections
```
mongodb-6b455c56fb-4lj56:2021-05-03T16:21:47.493+0000 I NETWORK  [conn1184792] end connection 10.247.145.228:40042 (2271 connections now open)
```

I was wondering what that number is on my system, so I `grep`ed my logs for `connections now open`, and this is what I got
```
2021-05-04T15:20:36.711+0000 I NETWORK  [conn1] end connection 127.0.0.1:36814 (0 connections now open)
2021-05-04T15:20:37.714+0000 I NETWORK  [conn2] end connection 127.0.0.1:36824 (0 connections now open)
2021-05-04T15:20:38.815+0000 I NETWORK  [conn3] end connection 127.0.0.1:36852 (0 connections now open)
2021-05-04T15:20:39.811+0000 I NETWORK  [conn4] end connection 127.0.0.1:36868 (0 connections now open)
2021-05-04T15:20:41.008+0000 I NETWORK  [conn5] end connection 127.0.0.1:36878 (0 connections now open)
```

I then started to kill the following pods:
* api-service
* shipyard-controller
* mongodb-datastore

and I finally got some numbers popping up:
```
2021-05-04T15:44:01.029+0000 I NETWORK  [conn6] end connection 10.60.1.6:42622 (54 connections now open)
2021-05-04T15:44:01.029+0000 I NETWORK  [conn15] end connection 10.60.1.6:43578 (52 connections now open)
2021-05-04T15:44:01.029+0000 I NETWORK  [conn19] end connection 10.60.1.6:43724 (51 connections now open)
2021-05-04T15:44:01.029+0000 I NETWORK  [conn17] end connection 10.60.1.6:43628 (55 connections now open)
2021-05-04T15:44:01.029+0000 I NETWORK  [conn5] end connection 10.60.1.6:42624 (50 connections now open)
2021-05-04T15:44:01.030+0000 I NETWORK  [conn14] end connection 10.60.1.6:43528 (49 connections now open)
2021-05-04T15:44:01.029+0000 I NETWORK  [conn16] end connection 10.60.1.6:43608 (53 connections now open)
2021-05-04T15:44:01.030+0000 I NETWORK  [conn18] end connection 10.60.1.6:43716 (48 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn38] end connection 10.60.3.7:59014 (47 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn1] end connection 10.60.3.7:51186 (27 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn36] end connection 10.60.3.7:59012 (45 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn35] end connection 10.60.3.7:56876 (46 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn30] end connection 10.60.3.7:53524 (44 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn27] end connection 10.60.3.7:52428 (43 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn22] end connection 10.60.3.7:52412 (42 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn25] end connection 10.60.3.7:52426 (41 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn28] end connection 10.60.3.7:52432 (40 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn31] end connection 10.60.3.7:53526 (38 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn26] end connection 10.60.3.7:52430 (39 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn37] end connection 10.60.3.7:59010 (36 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn23] end connection 10.60.3.7:52422 (37 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn20] end connection 10.60.3.7:52408 (35 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn21] end connection 10.60.3.7:52406 (34 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn24] end connection 10.60.3.7:52424 (33 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn29] end connection 10.60.3.7:53522 (32 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn7] end connection 10.60.3.7:51222 (31 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn4] end connection 10.60.3.7:51204 (30 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn3] end connection 10.60.3.7:51202 (29 connections now open)
2021-05-04T15:44:07.789+0000 I NETWORK  [conn2] end connection 10.60.3.7:51188 (28 connections now open)
```

This is now looking much more like there are some open connections that are never closed, which would lead to the aforementioned error on mongodb.

**Comment by christian-kreuzberger-dtx** [comment]
After killing remediation-service it seems that a lot of connections were closed. I will monitor that situation.

**Comment by christian-kreuzberger-dtx** [comment]
Still not sure if remedation-service is where the problem comes from, but there sure as hell is a memory-leak / connection-leak for MongoDB in remediation-service.

The connection-leak can be seen by looking at the logs of remediation-service and mongodb, the memory leak can be seen by looking at goroutine count:
![image](https://user-images.githubusercontent.com/56065213/117105573-9dceea80-ad7e-11eb-9997-15c9de74362e.png)

**Comment by christian-kreuzberger-dtx** [comment]
One of the pods must be starting to behave weird now:
![image](https://user-images.githubusercontent.com/56065213/117113679-1d62b680-ad8b-11eb-9a12-0f20ca435cd5.png)

**Comment by christian-kreuzberger-dtx** [comment]
@donovanmuller I am unable to reproduce the exact error that you are reporting.
```
SCRAM-SHA-1 authentication failed, storedKey mismatch
```
I believe that mongodb is crashing (which I was able to reproduce), and the error you are seeing is just a symptom of the crash.

One thing that could verify this is if you do the following:
```bash
$ kubectl -n keptn describe pod mongodb-6967b65894-hb6c5 # replace with your pod name
```
and look for container infos, e.g.:
```
Containers:
  mongodb:
    Container ID:   docker://ab61238b46c08f2762589d2279296fbd42ed6e2b36ea6a107a03bec3b74ab360
    Image:          docker.io/centos/mongodb-36-centos7:1
    Image ID:       docker-pullable://centos/mongodb-36-centos7@sha256:c93e5ea16ba6c2637e16d753a4c4128a99a09ee23a7c338694132eb2ca46d9ad
    Port:           27017/TCP
    Host Port:      0/TCP
    State:          Running
      Started:      Wed, 05 May 2021 10:52:27 +0200
    Last State:     Terminated
      Reason:       OOMKilled
      Exit Code:    137
      Started:      Wed, 05 May 2021 10:41:18 +0200
      Finished:     Wed, 05 May 2021 10:52:25 +0200
    Ready:          True
    Restart Count:  9
```
Here I can see that it was last killed because of OOMKilled, and that it has 9 restarts.

**Comment by donovanmuller** [comment]
@christian-kreuzberger-dtx picked this up:

```console
$ kubectl get events
LAST SEEN   TYPE      REASON                                                                                                        OBJECT                                               MESSAGE
59m         Warning   SystemOOM                                                                                                     node/ip-10-247-148-119.af-south-1.compute.internal   System OOM encountered, victim process: mongodb-datasto, pid: 16659
34m         Normal    ScaleDown                                                                                                     node/ip-10-247-148-119.af-south-1.compute.internal   marked the node as toBeDeleted/unschedulable
34m         Normal    Starting                                                                                                      node/ip-10-247-148-119.af-south-1.compute.internal   Starting kube-proxy.
33m         Normal    ClusterAutoscalerCleanup                                                                                      node/ip-10-247-148-119.af-south-1.compute.internal   removed ToBeDeletedTaint taint from node ip-10-247-148-119.af-south-1.compute.internal
23m         Normal    ScaleDown                                                                                                     node/ip-10-247-148-119.af-south-1.compute.internal   node removed by cluster autoscaler
23m         Normal    NodeNotReady                                                                                                  node/ip-10-247-148-119.af-south-1.compute.internal   Node ip-10-247-148-119.af-south-1.compute.internal status is now: NodeNotReady
22m         Normal    Deleting node ip-10-247-148-119.af-south-1.compute.internal because it does not exist in the cloud provider   node/ip-10-247-148-119.af-south-1.compute.internal   Node ip-10-247-148-119.af-south-1.compute.internal event: DeletingNode
22m         Normal    RemovingNode                                                                                                  node/ip-10-247-148-119.af-south-1.compute.internal   Node ip-10-247-148-119.af-south-1.compute.internal event: Removing Node ip-10-247-148-119.af-south-1.compute.internal from Controller
47m         Normal    Starting                                                                                                      node/ip-10-247-149-105.af-south-1.compute.internal   Starting kubelet.
47m         Normal    NodeHasSufficientMemory                                                                                       node/ip-10-247-149-105.af-south-1.compute.internal   Node ip-10-247-149-105.af-south-1.compute.internal status is now: NodeHasSufficientMemory
47m         Normal    NodeHasNoDiskPressure                                                                                         node/ip-10-247-149-105.af-south-1.compute.internal   Node ip-10-247-149-105.af-south-1.compute.internal status is now: NodeHasNoDiskPressure
47m         Normal    NodeHasSufficientPID                                                                                          node/ip-10-247-149-105.af-south-1.compute.internal   Node ip-10-247-149-105.af-south-1.compute.internal status is now: NodeHasSufficientPID
47m         Normal    NodeAllocatableEnforced                                                                                       node/ip-10-247-149-105.af-south-1.compute.internal   Updated Node Allocatable limit across pods
47m         Normal    RegisteredNode                                                                                                node/ip-10-247-149-105.af-south-1.compute.internal   Node ip-10-247-149-105.af-south-1.compute.internal event: Registered Node ip-10-247-149-105.af-south-1.compute.internal in Controller
47m         Normal    Starting                                                                                                      node/ip-10-247-149-105.af-south-1.compute.internal   Starting kube-proxy.
47m         Normal    NodeReady                                                                                                     node/ip-10-247-149-105.af-south-1.compute.internal   Node ip-10-247-149-105.af-south-1.compute.internal status is now: NodeReady
****
```

more specifically:

```console
$ kubectl get events | grep OOM
LAST SEEN   TYPE      REASON                                                                                                        OBJECT                                               MESSAGE
59m         Warning   SystemOOM                                                                                                     node/ip-10-247-148-119.af-south-1.compute.internal   System OOM encountered, victim process: mongodb-datasto, pid: 16659
```

**Comment by christian-kreuzberger-dtx** [comment]
That looks like mongodb-datastore (not mongodb) is getting an OOM.

It's impossible to say what's causing it, my tests so far haven't caused any problem with mongodb-datastore itself (only with mongodb directly).
Are you running certain API calls quite often or are you using a 3rd-party integration that queries a certain endpoint from mongodb-datastore? What kind of sequences are you running? How often?

**Comment by donovanmuller** [comment]
@christian-kreuzberger-dtx unfortunately don't have the time right now to dig in deeper but all I can say is that a cluster scaling event seems to trigger the behaviour in most cases. I.e. whenever a node is added/removed and the `mongo` Pods are moved around, this issue comes up for me.

**Comment by christian-kreuzberger-dtx** [comment]
We will keep an eye on this, but so far we haven't been able to reproduce it. I will close the issue for now.

**Comment by donovanmuller** [comment]
As a follow up to this issue. I kept experiencing the same issue as described initially. In my case it was always seemed to be related to a cluster autoscaling event where the Keptn MongoDB Pod would be rescheduled into another node etc.

I Googled a bit more and found this specific https://github.com/bitnami/bitnami-docker-mongodb/issues/159 on the `bitnami-docker-mongodb` image. It seemed very related to the issue I was experiencing, so I decided to replace the Keptn provided Mongo Helm Chart with the [Bitnami Chart](https://artifacthub.io/packages/helm/bitnami/mongodb).

Suffice to say I have not had this issue since deploying the Bitnami MongoDB Chart. Not sure if there is any further action but leaving this comment here for posterity.
